File: 백준/Silver/1913. 달팽이/달팽이.py
#달팽이

import sys
n=int(sys.stdin.readline()) #홀수 자연수 n
k=int(sys.stdin.readline())

#가운데 점을 찾고, 중앙부터, 방향: 위, 오,아래,왼, 위, 방향으로 탐색
#3->1,1 / 5->2,2
cen=n//2 #2
mapp=[[0]*n for _ in range(n)]

dx=[0,1,0,-1]  #한줄씩 증가하면 2씩증가...하는 맵,,(오, 아, 왼, 위)
dy=[1,0,-1,0]

x=y=cen
num=1
mapp[x][y]=num
ln=0
ans=[]
if num==k:
    ans=[x+1,y+1]

while True: 
    for dir in range(4):
        for _ in range(ln): #이동하는 걸 0번반복,, 2번반복,,,4번반복,,,
            x+=dx[dir] #해당 줄을 4방 탐색함
            y+=dy[dir]
            num+=1
            mapp[x][y]=num
            if num==k:
                ans=[x+1,y+1]
    if x==0 and y==0: #4번을 다 돌았는데, (0,0)까지 왔다면 break
        break
    x-=1 
    y-=1
    ln+=2

for i in range(n):
    print(*mapp[i])
print(*ans)    
    
    

File: 백준/Silver/1913. 달팽이/달팽이.java

import java.io.*;
import java.util.Arrays;


public class Main {
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		int n = Integer.parseInt(br.readLine());
		int ans = Integer.parseInt(br.readLine());
		int [][] mapp = new int[n][n];
		// 방향은 아래, 오른, 위, 왼 순으로..
		int dir = 0;
		int[] dx = {1,0,-1,0};
		int[] dy = {0,1,0,-1};
		int x=0;
		int y=0;
		int num=n*n;
		int ans_x=0;
		int ans_y=0;
		
		while (true) {
			mapp[x][y]=num;
			int nx=x+dx[dir];
			int ny=y+dy[dir];
			
			if (num==ans) {
				ans_x=x+1;
				ans_y=y+1;
			}
			
			if (nx<0 || ny<0 || nx>=n || ny>=n || mapp[nx][ny]!=0) {
				dir=(dir+1)%4;
			}
			
			
			x+=dx[dir];
			y+=dy[dir];
			num--;
			
			
			
		if (num==0) break;
		}
		
		
		
	for (int i=0; i<n; i++) {
		for (int j=0; j<n; j++) {
			sb.append(mapp[i][j]).append(" ");
		}
		System.out.println(sb);
		sb.setLength(0);
	}
	System.out.print(ans_x+" "+ans_y);
	}	
}


	